components/ratgdo/ratgdo.cpp

```diff
@@ -157,31 +170,81 @@ namespace ratgdo {
         byte2 = (data >> 24) & 0xff;
 
-        ESP_LOGV(TAG, "cmd=%03x (%s) byte2=%02x byte1=%02x nibble=%01x", cmd, cmd_name(cmd), byte2, byte1, nibble);
+        ESP_LOGD(TAG, "cmd=%03x (%s) byte2=%02x byte1=%02x nibble=%01x", cmd, cmd_name(cmd), byte2, byte1, nibble);
 
         if (cmd == command::STATUS) {
-            auto doorState = static_cast<DoorState>(nibble);
-            if (doorState == DoorState::DOOR_STATE_CLOSED && this->doorState != doorState) {
-                transmit(command::GET_OPENINGS);
+
+            this->doorState = static_cast<DoorState>(nibble);
+
+            if (this->doorState == DoorState::DOOR_STATE_OPENING && this->previousDoorState==DoorState::DOOR_STATE_CLOSED) {
+                this->startOpening = millis();
+            }
+            if (this->doorState == DoorState::DOOR_STATE_OPEN && this->previousDoorState==DoorState::DOOR_STATE_OPENING) {
+                if (this->startOpening > 0) {
+                    auto duration = (millis() - this->startOpening)/1000;
+                    duration = this->openingDuration > 0 ? (duration + this->openingDuration)/2 : duration;
+                    this->setOpeningDuration(round(duration*10)/10);
+                }
+            }
+            if (this->doorState == DoorState::DOOR_STATE_CLOSING && this->previousDoorState==DoorState::DOOR_STATE_OPEN) {
+                this->startClosing = millis();
+            }
+            if (this->doorState == DoorState::DOOR_STATE_CLOSED && this->previousDoorState==DoorState::DOOR_STATE_CLOSING) {
+                if (this->startClosing > 0) {
+                    auto duration = (millis() - this->startClosing)/1000;
+                    duration = this->closingDuration > 0 ? (duration + this->closingDuration)/2 : duration;
+                    this->setClosingDuration(round(duration*10)/10);
+                }
+            }
+            if (this->doorState == DoorState::DOOR_STATE_STOPPED) {
+                this->startOpening = -1;
+                this->startClosing = -1;
+            }
+
+            if (this->doorState == DoorState::DOOR_STATE_OPEN) {
+                this->doorPosition = 1.0;
+            } else if (this->doorState == DoorState::DOOR_STATE_CLOSED) {
+                this->doorPosition = 0.0;
+            } else {
+                if (this->closingDuration == 0 || this->openingDuration == 0 || this->doorPosition == DOOR_POSITION_UNKNOWN) {
+                    this->doorPosition = 0.5; // best guess
+                }
+            }
+
+            if (this->doorState == DoorState::DOOR_STATE_OPENING && !this->movingToPosition) {
+                this->positionSyncWhileOpening(1.0 - this->doorPosition);
+                this->movingToPosition = true;
+            }
+            if (this->doorState == DoorState::DOOR_STATE_CLOSING && !this->movingToPosition) {
+                this->positionSyncWhileClosing(this->doorPosition);
+                this->movingToPosition = true;
+            }
```

Comment:
```suggestion
            if (!this->movingToPosition) {
                if (this->doorState == DoorState::DOOR_STATE_OPENING) {
                    this->positionSyncWhileOpening(1.0 - this->doorPosition);
                } else {
                    this->positionSyncWhileClosing(this->doorPosition);
                }
                this->movingToPosition = true;
            }
```

one less conditional but doesn't really save that much